### PR TITLE
[SYCL-MLIR] SYCL 'accessor' analysis

### PR DESCRIPTION
**Pull request by sommerlukas**
MLIR analysis to infer properties of interest about instances of `sycl:accessor` from their construction in host code.

This also includes the ability to query two accessor for their potential aliasing behavior.

### CODE REVIEW

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.h [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+#include "mlir/Pass/AnalysisManager.h"
+
+namespace mlir {
+namespace polygeist {
+
+/// Represents information about a `sycl::accessor` gathered from its
+/// construction.
+class AccessorInformation {
+public:
+  AccessorInformation() {}
```
Comment (etiotto) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.h:32]: = default ?

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.h [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+  /// Returns the underlying buffer of this accessor.
+  Value getBuffer() const { return buffer; }
+
+  /// Returns true if buffer information is available for the buffer underlying
+  /// this accessor.
+  bool hasBufferInformation() const { return bufferInfo.has_value(); }
+
+  /// Returns the buffer information for the buffer underlying this accessor.
+  const BufferInformation &getBufferInfo() const { return *bufferInfo; }
```
Comment (etiotto) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.h:52]: assert that `hasBufferInformation()`

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.h [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                      bool hasOffset, llvm::ArrayRef<size_t> constOffset)
+      : buffer{buf}, bufferInfo{bufInfo}, needRange{hasRange},
+        constantRange{constRange}, needOffset{hasOffset},
+        constantOffset{constOffset} {}
+
+  /// Returns true if the underlying buffer of this accessor is known.
+  bool hasKnownBuffer() const { return buffer != nullptr; }
+
+  /// Returns the underlying buffer of this accessor.
+  Value getBuffer() const { return buffer; }
```
Comment (etiotto) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.h:45]: assert that `hasKnownBuffer()`

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+}
+
+AliasResult AccessorInformation::alias(const AccessorInformation &other,
+                                       AliasAnalysis &aliasAnalysis) const {
+  // If we can't analyze the underlying buffer, can't determine aliasing, so
+  // must assume they may alias.
+  if (!hasKnownBuffer() || !other.hasKnownBuffer())
+    return AliasResult::MayAlias;
+
+  auto isSameValue = [&](Value one, Value two) -> bool {
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp:111]: ### 
```suggestion
  bool isSameValue = [&](Value one, Value two) {
```

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+      if (hasConstantRange() && getBufferInfo().hasConstantSize() &&
+          SmallVector<size_t, 3>(getConstantRange()) ==
+              SmallVector<size_t, 3>(getBufferInfo().getConstantSize()) &&
+          !needsOffset())
+        // This accessor also covers the entire buffer, so they must alias
+        return AliasResult::MustAlias;
+
+      // This accessor only covers part of the buffer, so they alias for
+      // some elements.
+      return AliasResult::MayAlias;
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp:159]: This code is similar to lines 134-146. Possible to create a utility function or lambda to common up the code ?

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp (hunk 4 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp:111]: type of `isSameValue` is a function type.

**Review by jopperm — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp (hunk 4 above)
Comment (sommerlukas) [reply polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp:111]: Yes, `isSameValue` is a lambda function, therefore I was explicit about its return type, but used the more concise `auto` for the type in the declaration.

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+      // overlap.
+      return AliasResult::NoAlias;
+
+    // Could not refine further, assume they overlap partially, so may alias.
+    return AliasResult::MayAlias;
+  }
+
+  if (aliasAnalysis.alias(getBuffer(), other.getBuffer()) !=
+      AliasResult::NoAlias)
+    return AliasResult::MayAlias;
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp:209]: True, even if `MayAlias && nooverlap`, we could have aliasing. LGTM :+1:

**Review by victor-eds — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp (hunk 4 above)
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp:111]: sorry missed that, of course !

**Review by etiotto — APPROVED** [review]

**Pull request merged**